### PR TITLE
[Carry 18926] Add network internal mode

### PR DESCRIPTION
**Pull request by calavera**
Carry #18926.

/cc @thaJeztah, @vdemeester 

Signed-off-by: Chun Chen ramichen@tencent.com
Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

File: docs/reference/api/docker_remote_api.md [hunk 1]
```diff
@@ -113,6 +113,7 @@ This section lists each version from latest to oldest.  Each listing includes a
 * `GET /networks` now supports filtering by `name`, `id` and `type`.
 * `POST /containers/create` now allows you to set the static IPv4 and/or IPv6 address for the container.
 * `POST /networks/(id)/connect` now allows you to set the static IPv4 and/or IPv6 address for the container.
+* `POST /networks/create` now supports restricting external access to the network by setting a `internal` field.
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api.md]: :metal:

**Comment by thaJeztah** [comment]
LGTM

File: docs/reference/api/docker_remote_api.md (hunk 1 above)
Comment (tiborvass) [inline docs/reference/api/docker_remote_api.md]: `an` internal field... but the CI is almost green lol

**Comment by tiborvass** [comment]
apart from small nit, LGTM

File: docs/reference/api/docker_remote_api.md (hunk 1 above)
Comment (tiborvass) [inline docs/reference/api/docker_remote_api.md]: actually this should be `the` internal field, as @calavera privately suggested to me

Comment (thaJeztah) [inline docs/reference/api/docker_remote_api.md]: lol, yes, I only looked at the missing `.`, haha

Comment (calavera) [inline docs/reference/api/docker_remote_api.md]: replaced with `the`, because it's a specific field, not any, like `a` or `an` might suggest.

**Comment by chenchun** [comment]
Thanks for carrying this.

**Comment by xiangpengzhao** [comment]
All the existing [OPTIONS] of `docker network create` command (and almost other docker commands) are in the format as `--option_name=option_value`, so should we give the option `--internal` the same format as `--internal=bool`?
 I think this will keep unify style of commands and help users to understand the option better. @calavera

**Comment by thaJeztah** [comment]
@xiangpengzhao we no longer show `=false` on boolean flags. See https://github.com/docker/docker/pull/18082

**Comment by xiangpengzhao** [comment]
Sorry for not noticing that issue:sweat_smile: 
@thaJeztah

**Comment by sdurrheimer** [comment]
ping @albers future `docker network create --internal` flag

File: docs/reference/commandline/network_create.md [hunk 2]
```diff
@@ -22,6 +22,7 @@ parent = "smn_cli"
     --ipam-driver=default    IP Address Management Driver
     -o --opt=map[]           Set custom network plugin options
     --subnet=[]              Subnet in CIDR format that represents a network segment
+    --internal               Restricts external access to the network
```
Comment (vdemeester) [inline docs/reference/commandline/network_create.md]: Should be up, just before `--ip-range`.

File: man/docker-network-create.1.md [hunk 3]
```diff
@@ -14,6 +14,7 @@ docker-network-create - create a new network
 [**--ipam-driver**=*default*]
 [**-o**|**--opt**=*map[]*]
 [**--subnet**=*[]*]
+[**--internal**]
```
Comment (vdemeester) [inline man/docker-network-create.1.md]: Same here, should be upwards, just before `--ip-range`.

File: man/docker-network-create.1.md [hunk 4]
```diff
@@ -145,5 +151,8 @@ Be sure that your subnetworks do not overlap. If they do, the network create fai
 **--subnet**=[]
   Subnet in CIDR format that represents a network segment
 
+**--internal**
```
Comment (vdemeester) [inline man/docker-network-create.1.md]: And here :angel:

**Comment by vdemeester** [comment]
Small nits that could be done before merging (that's why I did put it back into docs review :stuck_out_tongue_closed_eyes:)

**Comment by thaJeztah** [comment]
good catch, @vdemeester

**Comment by icecrime** [comment]
Also needs a rebase :cat:

**Comment by calavera** [comment]
fixed comments and rebased

**Comment by vdemeester** [comment]
docs LGTM 🐹

**Comment by calavera** [comment]
all :green_heart: :tada:

**Comment by alvinr** [comment]
Not clear if this was intentional, but with `--internal` specified then the ports exposed do do appear in `docker inspect`. This breaks `docker ps` which shows empty values and utilities like `interlock` for HAproxy configuration.

Without `--internal` then you see the following stanza. This is excluded with the `--internal` flag.
```
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "45f20f5d699fe7fa7ea5ac95217ebea5b3cc47fd62202d8f430646e6d43e2a00",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "5000/tcp": [
                    {
                        "HostIp": "172.16.145.155",
                        "HostPort": "32768"
                    }
                ]
            },
```

**Pull request merged**